File: src/main/process-manager.ts
import { spawn, ChildProcess, execSync } from 'child_process'
import { IpcResult } from './types'

export class ProcessManager {
  private process: ChildProcess | null = null
  private currentHost: string = ''
  private onStreamEnd: ((code: number) => void) | null = null

  get isStreaming(): boolean {
    return this.process !== null
  }

  findBinary(): string | null {
    for (const name of ['chiaki', 'chiaki-ng']) {
      try {
        const result = execSync(`which ${name}`, { encoding: 'utf-8' }).trim()
        if (result) return result
      } catch { /* not found */ }
    }
    return null
  }

  startStream(
    nickname: string,
    host: string,
    options: { fullscreen?: boolean; dualsense?: boolean; passcode?: string } = {},
    onEnd: (code: number) => void
  ): IpcResult {
    if (this.isStreaming) return { success: false, error: 'A stream is already active' }

    const binary = this.findBinary()
    if (!binary) return { success: false, error: 'chiaki/chiaki-ng not found in PATH' }

    const args = ['stream', nickname, host]
    if (options.fullscreen) args.push('--fullscreen')
    if (options.dualsense) args.push('--dualsense')
    if (options.passcode) args.push('--passcode', options.passcode)

    this.currentHost = host
    this.onStreamEnd = onEnd

    try {
      this.process = spawn(binary, args, { stdio: 'inherit' })
    } catch (err) {
      this.process = null
      this.currentHost = ''
      return { success: false, error: `Failed to launch ${binary}: ${err instanceof Error ? err.message : String(err)}` }
    }

    this.process.on('error', (err) => {
      this.process = null
      this.currentHost = ''
      this.onStreamEnd?.(-1)
    })

    this.process.on('close', (code) => {
      this.process = null
      this.currentHost = ''
      this.onStreamEnd?.(code ?? 1)
    })

    return { success: true }
  }

  wakeUp(nickname: string, host: string, registKey: string): Promise<IpcResult> {
    return new Promise((resolve) => {
      const binary = this.findBinary()
      if (!binary) {
        resolve({ success: false, error: 'chiaki/chiaki-ng not found in PATH' })
        return
      }

      const args = ['wakeup', '--host', host, '--registkey', registKey]
      let stdout = ''
      let stderr = ''

      const proc = spawn(binary, args)
      proc.stdout?.on('data', (data: Buffer) => { stdout += data.toString() })
      proc.stderr?.on('data', (data: Buffer) => { stderr += data.toString() })

      proc.on('error', (err) => {
        resolve({ success: false, error: `Failed to launch ${binary}: ${err.message}` })
      })

      proc.on('close', (code) => {
        if (code === 0) {
          const message = stdout.trim() || 'Wake signal sent'
          resolve({ success: true, message })
        } else {
          const msg = stderr.trim() || `Process exited with code ${code}`
          resolve({ success: false, error: msg })
        }
      })
    })
  }

  stopStream(): void {
    this.process?.kill('SIGTERM')
  }
}

File: src/processmanager.cpp
#include "processmanager.h"
#include <QStandardPaths>

ProcessManager::ProcessManager(QObject *parent)
    : QObject(parent)
{
}

bool ProcessManager::isStreaming() const
{
    return m_process && m_process->state() == QProcess::Running;
}

QString ProcessManager::currentHost() const
{
    return m_currentHost;
}

QString ProcessManager::findChiakiBinary() const
{
    const QString path = QStandardPaths::findExecutable("chiaki");
    if (!path.isEmpty())
        return path;

    const QString ngPath = QStandardPaths::findExecutable("chiaki-ng");
    if (!ngPath.isEmpty())
        return ngPath;

    return QStringLiteral("chiaki");
}

void ProcessManager::startStream(const QString &nickname, const QString &host,
                                  bool fullscreen, bool dualsense, const QString &passcode)
{
    if (isStreaming())
        return;

    const QString binary = findChiakiBinary();

    QStringList args;
    args << "stream" << nickname << host;

    if (fullscreen)
        args << "--fullscreen";

    if (dualsense)
        args << "--dualsense";

    if (!passcode.isEmpty())
        args << "--passcode" << passcode;

    m_currentHost = host;
    m_process = new QProcess(this);

    connect(m_process, &QProcess::finished, this, &ProcessManager::onProcessFinished);

    m_process->start(binary, args);
    emit streamStarted();
    emit streamingChanged();
}

void ProcessManager::wakeUp(const QString &host, const QString &registKey)
{
    const QString binary = findChiakiBinary();

    QStringList args;
    args << "wakeup" << "--host" << host;

    if (!registKey.isEmpty())
        args << "--registkey" << registKey;

    auto *proc = new QProcess(this);
    connect(proc, &QProcess::finished, proc, &QProcess::deleteLater);
    proc->start(binary, args);
}

void ProcessManager::stopStream()
{
    if (!m_process)
        return;

    m_process->terminate();
}

void ProcessManager::onProcessFinished(int exitCode, QProcess::ExitStatus exitStatus)
{
    Q_UNUSED(exitStatus)

    m_process->deleteLater();
    m_process = nullptr;
    m_currentHost.clear();

    emit streamEnded(exitCode);
    emit streamingChanged();
}
